components/homme/src/theta-l_kokkos/cxx/CaarFunctorImpl.cpp

```diff
@@ -0,0 +1,638 @@
+#include "CaarFunctorImpl.hpp"
+
+namespace Homme {
+
+template <bool HYDROSTATIC, bool CONSERVATIVE>
+void CaarFunctorImpl::epoch1_blockOps()
+{
+  auto buffers_dp_tens = viewAsReal(m_buffers.dp_tens);
+  auto buffers_exner = viewAsReal(m_buffers.exner);
+  auto buffers_phi = viewAsReal(m_buffers.phi);
+  auto buffers_pnh = viewAsReal(m_buffers.pnh);
+  auto buffers_theta_tens = viewAsReal(m_buffers.theta_tens);
+  auto buffers_vdp = viewAsReal(m_buffers.vdp);
+
+  const Real data_eta_ave_w = m_data.eta_ave_w;
+  const int data_n0 = m_data.n0;
+
+  auto derived_vn0 = viewAsReal(m_derived.m_vn0);
+
+  const SphereGlobal sg(m_sphere_ops);
+
+  auto state_dp3d = viewAsReal(m_state.m_dp3d);
+  auto state_phinh_i = viewAsReal(m_state.m_phinh_i);
+  auto state_v = viewAsReal(m_state.m_v);
+  auto state_vtheta_dp= viewAsReal(m_state.m_vtheta_dp);
+
+  Kokkos::parallel_for(
+    "__PRETTY_FUNCTION__",
+    SphereBlockOps::policy(m_num_elems, 4),
+    KOKKOS_LAMBDA(const Team &team) {
+
+      SphereBlockOps b(sg, team);
+      if (b.skip()) return;
+
+      if (!HYDROSTATIC) {
+
+        const Real dphi = state_phinh_i(b.e,data_n0,b.x,b.y,b.z+1) - state_phinh_i(b.e,data_n0,b.x,b.y,b.z);
+        const Real vtheta_dp = state_vtheta_dp(b.e,data_n0,b.x,b.y,b.z);
+        if ((vtheta_dp < 0) || (dphi > 0)) abort();
+        EquationOfState::compute_pnh_and_exner(vtheta_dp, dphi, buffers_pnh(b.e,b.x,b.y,b.z), buffers_exner(b.e,b.x,b.y,b.z));
+
+        buffers_phi(b.e,b.x,b.y,b.z) = 0.5 * (state_phinh_i(b.e,data_n0,b.x,b.y,b.z) + state_phinh_i(b.e,data_n0,b.x,b.y,b.z+1));
+      }
+
+      const Real v0 = state_v(b.e,data_n0,0,b.x,b.y,b.z) * state_dp3d(b.e,data_n0,b.x,b.y,b.z);
+      const Real v1 = state_v(b.e,data_n0,1,b.x,b.y,b.z) * state_dp3d(b.e,data_n0,b.x,b.y,b.z);
+      buffers_vdp(b.e,0,b.x,b.y,b.z) = v0;
```

Comment:
The buffers are allocated based on the number of concurrent slots needed, but with SphereBlockOps (as well as with SphereColOps and SphereScanOps) we access them with the element index. We need to polish this out, so that we can use "team_idx" instead of the element index for the buffers.

Reply:
I use buffers to store values across kernels, so I need separate values for every element. I think you are saying that the buffers are actually allocated based on the number of teams, and the number of teams being the same as the number of local elements is a coincidence that I should not rely upon?

Reply:
Ah, I did not notice that the epochs are now separate kernels (rather than separate parts of the same kernel). That's a big change. It reminds me of what we did in SCREAM for some physics parametrizations, where an implementation with many smaller kernels was better than a monolithic one. But we should definitely see if this impl is also good for CUDA.

That said, yes, the buffers are allocated based on the maximum number of concurrent teams. If the number of (rank-local) elements is large enough, even on GPU we can get num_buffer_slots<num_elements, which means those buffers may be smaller than num_elems. This is pretty much always the case on CPU, btw, so this impl needs to be adjusted to be portable.

Reply:
How about the following to-do for me? Convert to using a separate set of buffers, created to be the required sizes.

Reply:
You mean make the new impl use different buffers from the original impl? That would work. But related to that, I am getting a bit concerned about how implementations for different backends may start do dramatically diverge. So yes, do the fix you suggest, but we need to start verification of other backends.